User's Templates (#17)
scramjetorg/scramjet-cloud-docs: templates/README.md

# Templates :pencil:

We have prepared for you some templates. User templates allow you to apply predefined settings to start building your own applications. Using templates, you can easily set up multiple applications with similar settings.

You can use it as a base for your own samples 👉 [sample template](https://github.com/scramjetorg/scramjet-cloud-docs/tree/main/templates). For this moment we support two variants of template in two programming languages:
 - JavaScript (Node.js) 👉 [template](https://github.com/scramjetorg/scramjet-cloud-docs/tree/main/templates/template-js)
 - TypeScript (ts-node) 👉 [template](https://github.com/scramjetorg/scramjet-cloud-docs/tree/main/templates/template-ts)

In the following sections we will describe both templates' contents and show you how they can be used in your own project.

## Work with JavaScript (Node.js)

This application package template contains two files:

> - package.json - JSON manifest file describing the app and its configuration such as main file to run. This is one of the Node.js project's basic requirement to have the `package.json` file in the root of every project. It contains metadata information in plain JSON Object format. This is the heart of any Node.js project It is used by the application to find its dependencies to install, scripts to run, etc. If you would like to start your project from scratch, without using our js-template, you would need to create a directory and run `npm init` inside. This command would create a `package.json` file for you with all the obligatory properties, for example:

Reviewer: I don't know why, but `JSON manifest file` reminds me of Progressive Web App where you can find manifest.json: https://web.dev/add-manifest/.

"The package.json file is kind of a manifest for your project." => JSON is like a ...?

Reviewer 2: An interesting observation. Now that I think about it, the manifest reminds me of the famous https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar :)

Author: fixed

  "main": "index.js",        // the entry point/file of the project
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"  // the script to run tests, you can add more scripts to this object
  },
  "author": "",              // the author of the project
  "license": "ISC"           // the license of the project
}
```

All the properties from the above are mandatory and some are not. Of course more properties can be added to this object if you need them, such as `dependencies` and `devDependencies`. Dependencies are simply all the other modules that your project will use. In template's `package.json` file there is one more property `"repository"`, which shows the repository address, where the source code of the package is located.

- index.js - this is where you should put your code and all the logic of the application you create. It will be the entry point of your application. In our template we introduce you to a very straight forward application, which simply reads input stream and write it to the output stream.

```js
const { PassThrough } = require("stream");

module.exports = function(input) {

>     // create clean output stream

Reviewer 2: ```suggestion
    // create a clean output stream
```

    const out = new PassThrough({encoding: 'utf-8'});

    input.on("data", data => {

>         // write some data to output stream

Reviewer 2: ```suggestion
       // write some data to the output stream
```

        out.write(data)
    });

>     // return output stream so it can be consumed (e.g. by CLI client)

Reviewer 2: ```suggestion
    // return the output stream so it can be consumed (e.g. by CLI client)
```

    return out;
};
```

> This is like the minimum signature for a function that takes a stream and returns a stream.

Reviewer 2: ```suggestion
This is the minimal signature for a function that takes a stream and returns a stream.
```

Of course more properties can be added to this object if you need them, such as `dependencies` and `devDependencies`. Dependencies are simply all the other modules that your project will use. In template's `package.json` file there is one more property `"repository"`, which shows the repository address, where the source code of the package is located.

> 🤓 If you would like to learn more about configuration in Node.js projects, please refer to the documentation on [Node.js official website](https://nodejs.dev/learn/introduction-to-nodejs).

Reviewer 2: ```suggestion
🤓 If you would like to learn more about the configuration in Node.js projects, please refer to the documentation on [Node.js official website](https://nodejs.dev/learn/introduction-to-nodejs).
```

}
```

> :nerd_face:  If you would like to learn more about configuration in TypeScript projects, please refer to the documentation on [TypeScript official website](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) and [ts-node official website](https://typestrong.org/ts-node/)..

Reviewer 2: ```suggestion
:nerd_face:  If you would like to learn more about the configuration in TypeScript projects, please refer to the documentation on [TypeScript official website](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) and [ts-node official website](https://typestrong.org/ts-node/).
```

In our template we introduce you to a very straight forward application, which simply reads input stream and write it to the output stream.

> :book: In the example below we use `@scramjet/types` module. Scramjet is a very nice and functional stream programming framework. You might find it very useful while writing your app or dealing with streams in general. More info, examples ang usage you can find [here](https://www.npmjs.com/package/scramjet).

```typescript
import { ReadableApp } from "@scramjet/types";
import { PassThrough } from "stream";

const app: ReadableApp<string> = async function(input) {
    /*
    THIS IS WHERE YOU SHOULD PUT YOUR CODE, AND ALL THE LOGIC OF YOUR APPLICATION

>     We put some code below for demonstration, and also instructions to run this template 

Reviewer 2: ```suggestion
    We put some code below for demonstration, and also instructions how to run this template 
```

To see how this template works you can run it with a few simple commands:

> > :bulb: **Note!** To run this template you need to have those two babies installed :dancers:: `npm install -g @scramjet/sth @scramjet/cli` 

Author: I am aware that it will not sound right but... `babies` are removed.
